iroh/src/commands/sync.rs

```diff
@@ -354,3 +480,273 @@ async fn print_entry(doc: &Doc, entry: &Entry, content: bool) -> anyhow::Result<
+        .to_vec();
+    let mut key = prefix.into_bytes().to_vec();
+    key.extend(suffix);
+    Ok(key)
+}
+
+#[derive(Debug, Clone)]
+struct ImportProgressState {
+    task: Arc<Mutex<Option<JoinHandle<()>>>>,
+    sender: flume::Sender<ImportProgressEvent>,
+}
+
+enum ImportProgressEvent {
+    Found { name: String, id: u64, size: u64 },
+    AddProgress { id: u64, progress: u64 },
+    AddingToDoc { id: u64 },
+    Done { id: u64 },
+    AllDone,
+    Error,
+}
+
+impl ImportProgressState {
+    fn new(doc_id: NamespaceId) -> Self {
+        let (sender, recv) = flume::bounded(32);
+        let task = tokio::spawn(async move {
+            let mp = MultiProgress::new();
+            let mut pbs = HashMap::new();
+            while let Ok(event) = recv.recv() {
+                match event {
+                    ImportProgressEvent::Found { name, id, size } => {
+                        let pb = mp.add(ProgressBar::new(size));
+                        pb.set_style(ProgressStyle::default_bar()
+            .template("{spinner:.green} [{bar:40.cyan/blue}] {msg} {bytes}/{total_bytes} ({bytes_per_sec}, eta {eta})").unwrap()
+            .progress_chars("=>-"));
+                        pb.set_message(name);
+                        pb.set_length(size);
+                        pb.set_position(0);
+                        pb.enable_steady_tick(Duration::from_millis(500));
+                        pbs.insert(id, pb);
+                    }
+                    ImportProgressEvent::AddProgress { id, progress } => {
+                        if let Some(pb) = pbs.get_mut(&id) {
+                            pb.set_position(progress);
+                        }
+                    }
+                    ImportProgressEvent::AddingToDoc { id } => {
+                        if let Some(pb) = pbs.get_mut(&id) {
+                            pb.finish_with_message(format!(
+                                "Adding to doc {}...",
+                                fmt_short(doc_id.as_bytes())
+                            ));
+                        }
+                    }
+                    ImportProgressEvent::Done { id } => {
+                        if let Some(pb) = pbs.remove(&id) {
+                            pb.finish_and_clear();
+                            mp.remove(&pb);
+                        }
+                    }
+                    ImportProgressEvent::AllDone | ImportProgressEvent::Error => {
+                        mp.clear().ok();
+                        return;
+                    }
+                }
+            }
+        });
+        Self {
+            task: Arc::new(Mutex::new(Some(task))),
+            sender: sender.clone(),
+        }
+    }
+
+    async fn found(&mut self, name: String, id: u64, size: u64) {
+        if !self.sender.is_disconnected() {
+            self.sender
+                .send_async(ImportProgressEvent::Found { name, id, size })
+                .await
+                .expect("checked");
```

Comment:
Why the check for is_disconnected?

Isn't this the same as `send_async(...).await.ok()`?

Reply:
ohh thank you yes that's better